### DISCUSSION TITLE
Complete state isolation between tests

### DISCUSSION THREAD

**Discussion by Dzeri96** [pytest-dev/pytest#12819] [Q&A] [answered]
_I know this question has been asked in similar forms many times before, but I was not able to find a solution to my exact problem, so please bear with me._

I'm trying to write tests for [batchframe](https://gitlab.com/Dzeri96/batchframe), which is a CLI tool that dynamically loads python modules based on user input, and it's causing issues in e2e tests where I want to simulate the full execution flow.

Let's assume I have two tests, where I point the tool to load two different modules from the project's `example` folder.

The problem comes from my usage of the `.__subclasses__()` function, which in the second test considers classes loaded in the first test.
As you know, fiddling with `sys.modules` is error-prone and does not even work in my case, because apparently the class lookup does not get cleared even when the parent module is unloaded. 

My question is, what would be the best way to go about testing such a CLI tool end-to-end?
I know I could launch a separate python process but I assume someone had this problem before and had found a better solution.

**Comment by RonnyPfannschmidt** [comment, marked as answer]
If the tool is implemented in a test hostile way, use subprocess, unless state can be fixed for the tool you stand no chance

**Comment by Dzeri96** [comment]
I'm sorry for the late follow-up, but does pytest have some facets/plugins to support this case? Perhaps you could point me to a project that has the same problem as mine?

**Reply by RonnyPfannschmidt** [reply]
the forked plugin is unmaintained and has issues, im not aware of a other isolation tool

**Comment by Dzeri96** [comment]
Alright. I'll close this issue then

**Reply by dyollb** [reply]
@Dzeri96 You could try https://github.com/dyollb/pytest-isolated (uses subprocess to isolate state)
